crates/amaru-ledger/src/rules/transaction/scripts.rs

```diff
@@ -1,59 +1,115 @@
 use std::collections::BTreeSet;
 
 use amaru_kernel::{
-    get_provided_scripts, HasScriptHash, MintedWitnessSet, ScriptHash, TransactionInput,
+    get_provided_scripts, DatumOption, HasAddress, HasDatum, HasScriptHash, Hash, MintedWitnessSet,
+    OriginalHash, PseudoScript, ScriptHash, ScriptRefWithHash, TransactionInput,
 };
 use thiserror::Error;
 
 use crate::context::{UtxoSlice, WitnessSlice};
 
 #[derive(Debug, Error)]
 pub enum InvalidScripts {
-    #[error("missing required scripts: missing {0:?}")]
+    #[error("missing required scripts: missing [{}]",
+        .0.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
     MissingRequiredScripts(Vec<ScriptHash>),
-    #[error("extraneous script witnesses: extra {0:?}")]
+    #[error("extraneous script witnesses: extra [{}]",
+        .0.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
     ExtraneousScriptWitnesses(Vec<ScriptHash>),
+    #[error("unspendable inputs; no datums: [{}]",
+        .0
+        .iter()
+        .map(|input|
+            format!("{}#{}", input.transaction_id, input.index)
+        )
+        .collect::<Vec<_>>()
+        .join(", ")
+    )]
+    UnspendableInputsNoDatums(Vec<TransactionInput>),
+    #[error(
+        "missing required datums: missing [{}] provided [{}]",
+        missing.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+        provided.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
+    MissingRequiredDatums {
+        missing: Vec<Hash<32>>,
+        provided: BTreeSet<Hash<32>>,
+    },
+    #[error(
+        "extraneous supplemental datums: allowed: [{}] provided [{}]",
+        allowed.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+        provided.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
+    ExtraneousSupplementalDatums {
+        allowed: BTreeSet<Hash<32>>,
+        provided: BTreeSet<Hash<32>>,
+    },
 }
 
+// TODO: this can be made MUCH more efficient. Remove clones, don't iterate the same list several times, etc... Lots of low hanging fruit.
 pub fn execute<C>(
     context: &mut C,
     reference_inputs: Option<&Vec<TransactionInput>>,
-    inputs: &Vec<TransactionInput>,
+    inputs: &[TransactionInput],
     witness_set: &MintedWitnessSet<'_>,
 ) -> Result<(), InvalidScripts>
 where
     C: UtxoSlice + WitnessSlice,
 {
     let required_scripts = context.required_scripts();
 
+    let resolved_inputs = inputs
+        .iter()
+        .filter_map(|input| {
+            context
+                // We assume that the input exists as that's validated during the inputs validation
+                .lookup(input)
+                .map(|output| (input, output))
+        })
+        .collect::<Vec<_>>();
+
+    let empty_vec = vec![];
+    let resolved_reference_inputs = reference_inputs
+        .unwrap_or(&empty_vec)
```

Comment:
Always hurt my eyes now I don't see how to improve this :D

Reply:
Use https://go.dev/